Add inverse kinematics module with interactive and benchmark examples

dylanturpin opened the pull request:
## Description
Adds a modular inverse kinematics (IK) system to Newton, inspired by [PyRoki](https://pyroki-toolkit.github.io/).

![output](https://github.com/user-attachments/assets/84df6811-6862-480b-ba36-d4671c609213)

**Core features:**
- Levenberg-Marquardt solver leveraging Warp's tile API for cholesky linear solve
(with adaptive damping, step accept/reject logic)
- Analytic Jacobians for speed (option to use autodiff or mix instead)
- Modular objectives: position, rotation, and joint limit objectives included

**API:**
```
import newton.core.ik as ik

ik_solver = ik.IKSolver(
    model=model,
    joint_q=joint_q, # (n_problems, n_coords) 
    objectives=[
        ik.PositionObjective(link_index=16, target_positions=targets, ...),
        ik.RotationObjective(link_index=16, target_rotations=rotations, ...),
        ik.JointLimitObjective(weight=0.1, ...)
    ],
    jacobian_mode=JacobianMode.ANALYTIC # or MIXED/AUTODIFF
)
ik_solver.solve(iterations=10)

```

**Interactive example:**

- `newton/examples/example_ik_interactive.py` - drag gizmos to control H1 robot end-effectors
- Supports "tied controls" mode where dragging one robot controls all environments
- GUI utilities in newton/examples/gizmo_utils.py

**Benchmark example:**

- `newton/examples/example_ik_benchmark.py --robot [franka/h1]`
- Runs many seeds and selects best
- Generates benchmark tables like below

## Benchmarks
Accuracy stats averaged over 2M samples.


### Franka (Simple Arm): 3x Faster, 99.994% Success Rate
Results (98th percentile errors for successful solves)

<table>
  <thead>
    <tr>
      <th rowspan="2">Solver</th>
      <th rowspan="2">Pos&nbsp;Err&nbsp;(mm)</th>
      <th rowspan="2">Ori&nbsp;Err&nbsp;(rad)</th>
      <th rowspan="2">Success&nbsp;(%)</th>
      <th colspan="5">time&nbsp;(ms)&nbsp;for&nbsp;batch&nbsp;size</th>
    </tr>
    <tr>
      <th>b&nbsp;= 1</th>
      <th>b&nbsp;= 10</th>
      <th>b&nbsp;= 100</th>
      <th>b&nbsp;= 1 k</th>
      <th>b&nbsp;= 2 k</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>cuRobo</td>
      <td>0.004482</td>
      <td>6.96e-6</td>
      <td>99.931</td>
      <td>7.952</td>
      <td>7.761</td>
      <td>11.811</td>
      <td>53.965</td>
      <td>85.420</td>
    </tr>
    <tr>
      <td>PyroKI</td>
      <td>0.000624</td>
      <td>6.21e-7</td>
      <td>99.865</td>
      <td>4.634</td>
      <td>5.275</td>
      <td>7.337</td>
      <td>34.851</td>
      <td>78.240</td>
    </tr>
    <tr>
      <td><strong>Newton</strong></td>
      <td><strong>0.000217</strong></td>
      <td><strong>2.60e-7</strong></td>
      <td><strong>99.994</strong></td>
      <td><strong>2.404</strong></td>
      <td><strong>0.977</strong></td>
      <td><strong>2.242</strong></td>
      <td><strong>13.588</strong></td>
      <td><strong>26.193</strong></td>
    </tr>
  </tbody>
</table>

### H1 (Humanoid): 10x Faster, 99.996% Success Rate
Note: Success requires ALL 4 end effectors to reach their targets
Results (98th percentile errors for successful solves)

<table>
  <thead>
    <tr>
      <th rowspan="2">Solver</th>
      <th rowspan="2">Pos&nbsp;Err&nbsp;(mm)</th>
      <th rowspan="2">Ori&nbsp;Err&nbsp;(rad)</th>
      <th rowspan="2">Success&nbsp;(%)</th>
      <th colspan="4">time&nbsp;(ms)&nbsp;for&nbsp;batch&nbsp;size</th>
    </tr>
    <tr>
      <th>b&nbsp;= 1</th>
      <th>b&nbsp;= 10</th>
      <th>b&nbsp;= 100</th>
      <th>b&nbsp;= 1&nbsp;k</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>PyroKI</td>
      <td>0.014312</td>
      <td>7.26e-6</td>
      <td>98.583</td>
      <td>11.599</td>
      <td>20.615</td>
      <td>108.481</td>
      <td>1525.460</td>
    </tr>
    <tr>
      <td><strong>Newton</strong></td>
      <td><strong>0.004949</strong></td>
      <td><strong>4.43e-7</strong></td>
      <td><strong>99.996</strong></td>
      <td><strong>1.809</strong></td>
      <td><strong>3.306</strong></td>
      <td><strong>23.344</strong></td>
      <td><strong>151.591</strong></td>
    </tr>
  </tbody>
</table>

## Limitations
- Analytic jacobians support only FIXED, REVOLUTE and PRISMATIC joints (more to be added soon)

Please ensure the migration guide for **warp.sim** users is up-to-date with the changes made in this MR.

- [x] The migration guide in ``docs/migration.rst`` is up-to date

## Before your PR is "Ready for review"

- [x] All commits are [signed-off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt--s) to indicate that your contribution adheres to the [Developer Certificate of Origin](https://developercertificate.org/) requirements
- [x] I understand that **GitHub** does not perform any GPU testing of this pull request
- [x] Necessary tests have been added
- [x] Documentation is up-to-date
- [x] Code passes formatting and linting checks with `pre-commit run -a`


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a high-performance, GPU-accelerated inverse kinematics (IK) solver supporting multiple Jacobian modes and batch problem solving.
  * Added example scripts for benchmarking IK performance and interactive IK manipulation with 3D visualization and multi-environment batching.
  * Provided configurable robot models, multi-seed initialization, and detailed IK error evaluation and reporting.
  * Enabled user-friendly 3D manipulation of robot end-effectors with optional tied or independent target control across environments.

* **Tests**
  * Added comprehensive unit tests verifying IK solver convergence, Jacobian correctness, and forward kinematics parity across joint types and devices.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dylanturpin commented on newton/examples/gizmo_utils.py:
```diff
@@ -0,0 +1,958 @@
+        if self.drag_state.mode == "translate" and self.position_callback:
+            if not np.array_equal(prev_position, self.drag_state.target.get_position()):
+                self.position_callback(self.drag_state.target.id, self.drag_state.target.get_position())
+            return self.drag_state.mode == "rotate"
```
The code is correct as written. There's an existing callback to pan the view when you click and drag. We still want this to fire when the user translates a gizmo (so view follows its movement). For rotation it feels awkward if the view changes while you drag, so we return True to prevent the pan fallback from firing.

shi-eric commented:
Running the tests individually on my system, `test_ik_fk_kernels.py` took 137 secs and `test_ik.py` took 128.4 secs (using the times reported by unittest).  The module load times and recompilations are a bit diabolical:

```
Module newton.sim.ik 54ecea2 load on device 'cpu' took 5404.95 ms  (compiled)
Module newton.sim.ik 44a6e7e load on device 'cpu' took 5191.23 ms  (compiled)
Module newton.sim.ik d7f2af7 load on device 'cpu' took 5328.65 ms  (compiled)
Module newton.sim.ik aef92ac load on device 'cpu' took 5585.66 ms  (compiled)
```

```
Module newton.sim.ik 183cd3f load on device 'cuda:0' took 105012.24 ms  (compiled)
```

```
Module newton.sim.ik fee9900 load on device 'cuda:0' took 38290.72 ms  (compiled)
Module newton.sim.ik d8bf09c load on device 'cuda:0' took 38255.77 ms  (compiled)
```

```
Module newton.sim.ik 2524012 load on device 'cuda:0' took 41002.01 ms  (compiled)
```

It would be helpful if you could take a look at `newton.sim.ik` and see what's causing these recompilations and how they can be eliminated.

dylanturpin commented:
So, my mistake appears to be that I keyed the cache on problem dimension only. It also needs to take CPU / CUDA into account.

So I'd compile once on CPU (for a given problem size), later run on GPU, see a (spurious) cache hit, lazy recompile to CUDA (maybe replacing the CPU cached version) and so on and so on.

Should be an easy fix. Can I just call wp.get_device().is_cuda?

shi-eric commented on newton/sim/ik.py:
```diff
@@ -0,0 +1,1596 @@
+    TILE_N_RESIDUALS = None
+    _cache: ClassVar[dict[tuple[int, int], type]] = {}
+
+    def __new__(cls, model, joint_q, objectives, *a, **kw):
```
...but then immediately after writing that I thought about the multi-GPU case with the same arch. So many just key off `wp.get_device().arch` if `is_cuda` is too broad.

dylanturpin replied:
So with this fix, I see correct compile behaviour (cache hits/misses & compilations happen only when they should). But I'm not seeing it affect wall clock speed locally. Is there a particular command I should run to repro the slownesss?

dylanturpin replied:
I was trying just my individual tests.

I can repro locally test_box_box_contact taking 20s on main vs. many minutes on this branch, even running it in isolation. That's pretty confusing.

shi-eric replied:
I assume your dependencies are up-to-date and this branch is up to date with main?

dylanturpin replied:
I think the pushed one here is 5 commits behind main. Locally I rebased, shall I push the rebased version?

The link to test_box_box_contact turns out to be non-mysterious. At the top of ik.py, I have...
```
# TODO: debug why this is necessary
wp.config.enable_backward = True
```

If I left that out, my ik lib would fail in autodiff mode.

This seems to be conflicting with
```
# TODO(team): improve compile time to enable backward pass
wp.config.enable_backward = False
```
at the top of `gjk.py`.

So... if those both end up getting imported... enable_backward will be on our off globally depending on order.

I believe `gjk.py` should make the change at the module level. And then I can delete my line.

I.e., in gjk.py... `wp.set_module_options({"enable_backward": False})`.

dylanturpin replied:
Oh... it's a bit more complex.

`gjk.py` only defines functions. So we don't know what module those functions will be compiled into (and which one's enable_backward they will respect) until someone uses them in a kernel. I'm guessing that's why they used the global setting.

dylanturpin replied:
For now I can leave `gjk.py` and use `wp.set_module_options({"enable_backward": True})` in ik.py.

Perhaps I should create an issue for this?

shi-eric replied:
Sounds good!

shi-eric replied:
Thank you! Can you also open up your newton-physics membership settings so I don't have to approve your workflows?

 1. Go to the organization's People page: https://github.com/orgs/newton-physics/people
2. Find your username in the list.
3. Click the dropdown next to your name and change your visibility from 'Private' to 'Public'.

dylanturpin replied:
Done!

eric-heiden commented on newton/examples/example_ik_benchmark.py:
```diff
@@ -0,0 +1,426 @@
+        self.cfg.parser(asset_path, builder, floating=False)
+
+        model = builder.finalize(requires_grad=False)
+        model.ground = True
```
You need to call `builder.add_ground_plane()` now, `model.ground` has been removed.

eric-heiden commented on newton/sim/ik.py:
```diff
@@ -0,0 +1,1597 @@
+        self.costs_proposed = wp.zeros_like(self.costs_current)
+        self.lambda_values = wp.zeros(self.n_problems, dtype=wp.float32)
+        self.accept_flags = wp.zeros(self.n_problems, dtype=wp.int32)
+        self.pred_reduction = wp.zeros(self.n_problems, dtype=wp.float32)
```
Make sure these Warp arrays are allocated on `model.device`.

eric-heiden commented on newton/sim/ik.py:
```diff
@@ -0,0 +1,1597 @@
+    ----------
+    model : newton.Model
+        Articulation describing joints/bodies.
+    joint_q : ndarray[float32] shape (n_problems, model.joint_coord_count)
```
Isn't `joint_q` a Warp array?

mmacklin replied:
Is this really `ndarray` or should it be `wp.array`?

eric-heiden commented on newton/sim/ik.py:
```diff
@@ -0,0 +1,1597 @@
+
+    Notes
+    ANALYTIC and MIXED modes currently only support models with revolute, prismatic, and fixed joints.
+    For more complex joint types (e.g., free, ball, universal, compound, D6), use AUTODIFF mode.
```
Joint types `universal` and `compound` don't exist anymore (D6 handles those).

eric-heiden commented on newton/sim/ik.py:
```diff
@@ -0,0 +1,1597 @@
+                ],
+                outputs=[
+                    self.joint_S_s,
+                ],
```
Make sure the `device` argument is set correctly (probably `self.model.device`).

eric-heiden commented on newton/sim/ik.py:
```diff
@@ -0,0 +1,1597 @@
+
+
+@wp.func
+def compute_3d_rotational_dofs_pos(
```
Is this function and also `compute_2d_rotational_dofs_pos` called anywhere?

eric-heiden commented on newton/sim/ik.py:
```diff
@@ -0,0 +1,1597 @@
+    cost = float(0.0)
+    for i in range(num_residuals):
+        r = residuals[problem_idx, i]
+        cost += r * r
```
Just a question: are these dynamic loops now fully differentiable? I lost track whether this has been fixed now in Warp @daedalus5, @christophercrouzet

daedalus5 replied:
No, they aren't in general. However, in this case it should be okay: there are no intermediate values from the forward pass required to compute any adjoints in the backward pass.

eric-heiden commented on newton/sim/ik.py:
```diff
@@ -0,0 +1,1597 @@
+            e = np.zeros((n_problems, total_residuals), dtype=np.float32)
+            for prob_idx in range(n_problems):
+                e[prob_idx, residual_offset + component] = 1.0
+            self.e_arrays.append(wp.array(e.flatten(), dtype=wp.float32))
```
Need to allocate on the correct device, probably `target_positions.device`.

eric-heiden commented on newton/tests/test_ik_fk_kernels.py:
```diff
@@ -0,0 +1,246 @@
+    return model
+
+
+def _randomise_joint_q(model: newton.Model, seed: int = 0) -> None:
```
American English spelling please :)

eric-heiden commented on newton/examples/example_ik_interactive.py:
```diff
@@ -0,0 +1,529 @@
+        for pos in env_offsets:
+            builder.add_builder(articulation_builder, xform=wp.transform(pos, upright_quat))
+
+        builder.up_axis = "Y"
```
Can you keep the original `up_axis` (Z)? I don't see why you need to change it.

mmacklin commented on newton/sim/__init__.py:
```diff
@@ -27,6 +27,14 @@
 )
 from .graph_coloring import color_graph, plot_graph
+from .ik import (
+    IK,
```
I don't think we should have these on the top level `newton.sim` namespace, especially for things like `JacobianMode`, `PositionObjective`, these are very general terms that could be related to anything in simulation.

I would suggest users to import directly the ik module e.g.: `import newton.sim.ik as ik` and then just use it directly.

mmacklin commented on newton/sim/ik.py:
```diff
@@ -0,0 +1,1597 @@
+        Ordered list of objectives whose residuals are concatenated.
+    lambda_initial : float, default 0.1
+        Initial LM damping per problem.
+    jacobian_mode : JacobianMode, default JacobianMode.AUTODIFF
```
Can we group related parameters together?

Not sure why `jacobian_mode` appears in between all the `lambda_` parameters but will be good if we can group these logically with all the same prefixes appearing together.

mmacklin commented on newton/sim/ik.py:
```diff
@@ -0,0 +1,1597 @@
+        Articulation describing joints/bodies.
+    joint_q : ndarray[float32] shape (n_problems, model.joint_coord_count)
+        Initial joint coordinates for every independent problem.
+    objectives : Sequence[IKObjective]
```
So I think we might want some more descriptive docs for the `objectives` in particular - e.g.: my first question was "are these specified per-problem, or per-model"? I think it would be worth clarifying here that IKObjectives can store a list of targets, one per-problem but objectives themselves are shared across problems.

dylanturpin commented on newton/sim/__init__.py:
```diff
@@ -27,6 +27,14 @@
 )
 from .graph_coloring import color_graph, plot_graph
+from .ik import (
+    IK,
```
@mmacklin @eric-heiden I was avoiding use of `solver` since it seems like solver = integrator in newton. That's why all the somewhat contrived ... `IK` and `IK_system` type naming.

Which matters more?
- Clarity of purpose of calling it `IKSolver` or
- avoiding mixed semantics for `Solver` in newton generally

dylanturpin replied:
I think clarity of `IKSolver` wins. I'll resolve for now.

dylanturpin commented:
Thanks to all for detailed comments!

Resolved:
* **Deprecated ground plane:** removed `model.ground`; now call `builder.add_ground_plane()` everywhere.
* **Gizmo helpers relocated:** moved to `newton/utils/gizmo.py`, deleted old file, updated imports.
* **Solver rename:** `IK` renamed to `IKSolver`; updated all imports, examples, tests and docs.
* **Namespace cleanup:** stopped reexporting IK symbols from `newton.sim.__init__`.
* **Device correctness:** every `wp.zeros`, `wp.array`, and `wp.launch` now passes `device=self.device`.
* **Documentation improvements:**
  - grouped `lambda_*` parameters together.
  - documented `joint_q` as `wp.array`.
  - removed obsolete joint types “universal” and “compound”.
  - added detailed problem batch‑structure explanation.
* **Dead code removed:** deleted `compute_2d_rotational_dofs_pos` and `compute_3d_rotational_dofs_pos`.
* **Tests updated:**
  - American spelling: `_randomize_joint_q`.
  - added non‑identity parent/child transforms for FK tests.
* **Examples adjusted:**
  - removed unnecessary `up_axis` setting from both examples
  
Let me know if there’s anything else to tweak, otherwise should be ready to merge.

dylanturpin commented on newton/examples/example_ik_benchmark.py:
```diff
@@ -0,0 +1,419 @@
+        self.results = []
+        self.rng = np.random.default_rng()
+
+    def _create_robot(self) -> newton.Model:
+        builder = newton.ModelBuilder()
+        builder.num_rigid_contacts_per_env = 0
+        builder.default_shape_cfg.density = 100.0
+
+        asset_path = (
+            Path(self.cfg.asset) / self.cfg.file
+            if ("/" in self.cfg.asset or "\\" in self.cfg.asset)
+            else newton.utils.download_asset(self.cfg.asset) / self.cfg.file
+        )
+        self.cfg.parser(asset_path, builder, floating=False)
+
+        model = builder.finalize(requires_grad=False)
+        return model
+
```
So, I think I'll follow the practice I see elsewhere in the library and leave it to just
```
        asset_path = newton.utils.download_asset(self.cfg.asset) / self.cfg.file
        self.cfg.parser(asset_path, builder, floating=False)
```

`download_asset` already has internal handling for caching. In terms of catching errors... if we decide that's a problem we should fix it in generality later, for now I'll follow standard from elsewhere in the library (and remove my own ad-hoc caching). Pushing a fix...

dylanturpin commented:
Now that PR #413 is in, deleted:

```
# TODO: Remove this once gjk.py no longer sets global option to False
wp.set_module_options({"enable_backward": True})
```

mzamoramora-nvidia commented:
Very cool work here. Could it be that gizmo.py is missing from newton.utils?

dylanturpin commented:
@mzamoramora-nvidia Thank you, good catch! Fixed in #423
